### ISSUE TITLE
Differences from original tensorflow/deepdream

### ISSUE THREAD

**Issue by genekogan** [ProGamerGov/Protobuf-Dreamer#3]
Just stumbled on this! It looks to me like the results produced by this repo, and @jnordberg's implementation, are much sharper and overall better-looking than the original [tensorflow deepdream](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/tutorials/deepdream). 

Comparing the codebases, it looks mostly the same, rewritten slightly, but I can't find major differences. I am wondering if there's anything I missed, an adaptation from the original, that contributes to the better sharpness/color balance.

**Comment by ProGamerGov** [comment]
@genekogan  I'm not 100% sure what could be causing the better sharpeness, but the original Tensorflow DeepDream code uses a deprocessing function while this project does not.

Also, this project uses:
```
img += g*(step / (np.abs(g).mean()+1e-7))
```

While the Tensorflow v1 DeepDream code uses this:

```
g /= g.std()+1e-8         # for different layers and networks
img += g*step
```

In my experience with PyTorch, `mean()`, `abs()`, `norm()`, `input.pow(2).sum().sqrt()`, and any combinations of those will produce difference results in terms of quality and content.

**Comment by genekogan** [comment]
Ah yes, I missed that. I also just noticed a difference in how the image is resized for each octave. The lines:

```
lo = resize(img, np.int32(np.float32(hw)/octave_scale))
hi = img-resize(lo, hw)
```

which is later added back:

```
img = resize(img, hi.shape[:2])+hi
```

Is `hi` attempting to correct for resize artifacts/distortion? The original doesn't do this, it just has `img = resize(img, np.int32(hw))` at each octave.

**Comment by ProGamerGov** [comment]
@genekogan Actually for the above code, I was comparing the `render_naive` and not the `render_deepdream` function from the original Tensorflow (and you may have been doing the same?).

```
def render_naive(t_obj, img0=img_noise, iter_n=20, step=1.0):
    t_score = tf.reduce_mean(t_obj) # defining the optimization objective
    t_grad = tf.gradients(t_score, t_input)[0] # behold the power of automatic differentiation!
    
    img = img0.copy()
    for i in range(iter_n):
        g, score = sess.run([t_grad, t_score], {t_input:img})
        # normalizing the gradient, so the same step size should work 
        g /= g.std()+1e-8         # for different layers and networks
        img += g*step
        print(score, end = ' ')
    clear_output()
    showarray(visstd(img))
```

```
def render_deepdream(t_obj, img0=img_noise,
                     iter_n=10, step=1.5, octave_n=4, octave_scale=1.4):
    t_score = tf.reduce_mean(t_obj) # defining the optimization objective
    t_grad = tf.gradients(t_score, t_input)[0] # behold the power of automatic differentiation!

    # split the image into a number of octaves
    img = img0
    octaves = []
    for i in range(octave_n-1):
        hw = img.shape[:2]
        lo = resize(img, np.int32(np.float32(hw)/octave_scale))
        hi = img-resize(lo, hw)
        img = lo
        octaves.append(hi)
    
    # generate details octave by octave
    for octave in range(octave_n):
        if octave>0:
            hi = octaves[-octave]
            img = resize(img, hi.shape[:2])+hi
        for i in range(iter_n):
            g = calc_grad_tiled(img, t_grad)
            img += g*(step / (np.abs(g).mean()+1e-7))
            print('.',end = ' ')
        clear_output()
        showarray(img/255.0)
```
The `render_deepdream` function also resizes images for octaves the same way.

https://github.com/tensorflow/examples/blob/master/community/en/r1/deepdream.ipynb

And this is the function from my modified version of @jnordberg's project:

```
def render_deepdream(t_grad, img0, iter_n=10, step=1.5, octave_n=4, octave_scale=1.4):
    # split the image into a number of octaves
    img = img0
    octaves = []
    for i in range(octave_n-1):
        hw = img.shape[:2]
        lo = resize(img, np.int32(np.float32(hw)/octave_scale))
        hi = img-resize(lo, hw)
        img = lo
        octaves.append(hi)

    # generate details octave by octave
    for octave in range(octave_n):
        if octave>0:
            hi = octaves[-octave]
            img = resize(img, hi.shape[:2])+hi
        for i in range(iter_n):
            #g = calc_grad_tiled(img, t_grad)
	    g = calc_grad_tiled(img, t_grad, tile_size)
            img += g*(step / (np.abs(g).mean()+1e-7))
	    if not verbose:
	        print "Iteration Number: %d" % i
	if not verbose:
            print "Octave Number: %d" % octave


    return Image.fromarray(np.uint8(np.clip(img/255.0, 0, 1)*255))
```

Also, have you ever heard of the #proj-lucid/distill.pub Slack channel/group before?

**Comment by genekogan** [comment]
Yes, I've used Lucid a bit, and hung out on their Slack. I tried to incorporate my masks into it at some point, but had some trouble and didn't get too far. Considering revisiting it at some point.

I'm trying to do some A vs B tests to compare. I'm wondering if you can share the parameters you used (i.e. step size, octave ratio, number of octaves, num iterations, etc) for the large images (e.g. [this one](https://raw.githubusercontent.com/ProGamerGov/Protobuf-Dreamer/master/examples/inception5h/mixed5c_pool_reduce_61.jpg)) on the readme page. You have size (2048, 2048) images which are quite intricate, but I'm having trouble finding the right numbers to reproduce them.

**Comment by ProGamerGov** [comment]
@genekogan I was able to find a bunch of parameter sets that I likely used to create the example images, but it's been over 3 years and I'm not exactly sure which is which.

For a bunch of things, I used `mixed4a_3x3_bottleneck_pre_relu:51` as the input image, and I would often denote the size of square image by adding 640, 4096, etc... to it's name.


I think I may have been testing what channels looked like here:


```
python pb_dreamer.py --input_image mixed4a_3x3_bottleneck_pre_relu_51_640.png --output_image mixed4c_3x3_bottleneck_channel_08.png --octaves 8 --iter 25 --layer mixed4c_3x3_bottleneck --channel 8

python pb_dreamer.py --input_image mixed4a_3x3_bottleneck_pre_relu_51_640.png --output_image mixed4c_3x3_bottleneck_channel_09.png --octaves 8 --iter 25 --layer mixed4c_3x3_bottleneck --channel 9

python pb_dreamer.py --input_image mixed4a_3x3_bottleneck_pre_relu_51_640.png --output_image mixed4c_3x3_bottleneck_channel_10.png --octaves 8 --iter 25 --layer mixed4c_3x3_bottleneck --channel 10
```

It also looks like I was running the outputs through the script again with the same parameters:

```
python pb_dreamer.py --input_image mixed4c_pool_reduce_channel_01.png --output_image part2/mixed4c_pool_reduce_2_channel_01.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 1

python pb_dreamer.py --input_image mixed4c_pool_reduce_channel_02.png --output_image part2/mixed4c_pool_reduce_2_channel_02.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 2

python pb_dreamer.py --input_image mixed4c_pool_reduce_channel_03.png --output_image part2/mixed4c_pool_reduce_2_channel_03.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 3

python pb_dreamer.py --input_image mixed4c_pool_reduce_channel_04.png --output_image part2/mixed4c_pool_reduce_2_channel_04.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 4

python pb_dreamer.py --input_image mixed4c_pool_reduce_channel_05.png --output_image part2/mixed4c_pool_reduce_2_channel_05.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 5

python pb_dreamer.py --input_image mixed4c_pool_reduce_channel_06.png --output_image part2/mixed4c_pool_reduce_2_channel_06.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 6
```


```
python pb_dreamer.py --input_image mixed4c_pool_reduce_2_channel_17_4096.png --output_image mixed4c_pool_reduce_2_channel_17_4096_2.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 17 --step_size 2

python pb_dreamer.py --input_image mixed4c_pool_reduce_2_channel_17_4096_2.png --output_image mixed4c_pool_reduce_2_channel_17_4096_2_2.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 17 --step_size 2

python pb_dreamer.py --input_image mixed4c_pool_reduce_2_channel_17_4096_2_2.png --output_image mixed4c_pool_reduce_2_channel_17_4096_2_2_2.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 17 --step_size 2

python pb_dreamer.py --input_image mixed4c_pool_reduce_2_channel_17_4096_2_2_2.png --output_image mixed4c_pool_reduce_2_channel_17_4096_2_2_2_2.png --octaves 8 --iter 25 --layer mixed4c_pool_reduce --channel 17 --step_size 2



python pb_dreamer.py --input_image mixed5a_1x1_2_channel_198_4096.png --output_image mixed5a_1x1_2_channel_198_4096_2.png --octaves 8 --iter 25 --layer mixed5a_1x1 --channel 198

python pb_dreamer.py --input_image mixed5a_1x1_2_channel_198_4096_2.png --output_image mixed5a_1x1_2_channel_198_4096_2_2.png --octaves 8 --iter 50 --layer mixed5a_1x1 --channel 198


python pb_dreamer.py --input_image mixed4e_2_channel_62_4096.png --output_image mixed4e_2_channel_62_4096_2.png --octaves 8 --iter 25 --layer mixed4e --channel 62 --step_size 2

python pb_dreamer.py --input_image mixed4e_2_channel_62_4096_2.png --output_image mixed4e_2_channel_62_4096_2_@.png --octaves 8 --iter 25 --layer mixed4e --channel 62 --step_size 2

python pb_dreamer.py --input_image mixed4e_2_channel_62_4096_2.png --output_image mixed4e_2_channel_62_4096_2_@.png --octaves 8 --iter 25 --layer mixed4e --channel 62 --step_size 2

python pb_dreamer.py --input_image mixed4e_2_channel_62_4096_2_@_6144.png --output_image mixed4e_2_channel_62_4096_2_@_6144_2.png --octaves 8 --iter 25 --layer mixed4e --channel 62 --step_size 2

python pb_dreamer.py --input_image mixed5a_1x1_2_channel_134_4096.png --output_image mixed5a_1x1_2_channel_134_4096_2.png --octaves 8 --iter 25 --layer mixed4e --channel 134 --step_size 2

python pb_dreamer.py --input_image mixed5a_1x1_2_channel_134_4096_2.png --output_image mixed5a_1x1_2_channel_134_4096_2_2.png --octaves 8 --iter 25 --layer mixed4e --channel 134 --step_size 2

python pb_dreamer.py --input_image mixed5a_1x1_2_channel_134_4096_2.png --output_image mixed5a_1x1_2_channel_134_4096_2_2.png --octaves 8 --iter 25 --layer mixed5a_1x1 --channel 134 --step_size 1.8

python pb_dreamer.py --input_image mixed5a_1x1_2_channel_134_4096_2_2.png --output_image mixed5a_1x1_2_channel_134_4096_2_2_2.png --octaves 8 --iter 25 --layer mixed5a_1x1 --channel 134 --step_size 1.9
```


It looks like this one was me experimenting with ImageMagick and multiscale generation:


```
convert mixed4c_2_channel_477.png -resize 768 mixed4c_2_channel_477_768.png

python pb_dreamer.py --input_image mixed4c_2_channel_477_768.png --output_image out_mixed4c_2_channel_477_768.png --octaves 8 --iter 25 --layer mixed4c --channel 477

convert out_mixed4c_2_channel_477_768.png -resize 1024 mixed4c_2_channel_477_1024.png

python pb_dreamer.py --input_image mixed4c_2_channel_477_1024.png --output_image out_mixed4c_2_channel_477_1024.png --octaves 8 --iter 25 --layer mixed4c --channel 477

convert out_mixed4c_2_channel_477_1024.png -resize 1536 mixed4c_2_channel_477_1536.png

python pb_dreamer.py --input_image mixed4c_2_channel_477_1536.png --output_image out_mixed4c_2_channel_477_1536.png --octaves 8 --iter 25 --layer mixed4c --channel 477

convert out_mixed4c_2_channel_477_1536.png -resize 2048 mixed4c_2_channel_477_2048.png

python pb_dreamer.py --input_image mixed4c_2_channel_477_2048.png --output_image out_mixed4c_2_channel_477_2048.png --octaves 8 --iter 25 --layer mixed4c --channel 477

convert out_mixed4c_2_channel_477_2048.png -resize 4096 mixed4c_2_channel_477_4096.png

python pb_dreamer.py --input_image mixed4c_2_channel_477_4096.png --output_image out_mixed4c_2_channel_477_4096.png --octaves 8 --iter 25 --layer mixed4c --channel 477

python pb_dreamer.py --input_image out_mixed4c_2_channel_477_4096.png --output_image out_mixed4c_2_channel_477_4096_2.png --verbose --octaves 8 --iter 25 --layer mixed4c --channel 477

--verbose

convert out_mixed4c_2_channel_477_4096_2.png -resize 8192 out_mixed4c_2_channel_477_8192.png

python pb_dreamer.py --input_image mixed4c_2_channel_477_8192.png --output_image out_mixed4c_2_channel_477_8192.png --verbose --octaves 8 --iter 25 --layer mixed4c --channel 477




python pb_dreamer.py --input_image mixed4c_2_channel_477_2.png --output_image mixed4c_2_channel_477_2_2.png --octaves 8 --iter 25 --layer mixed4c --channel 477

python pb_dreamer.py --input_image mixed4c_2_channel_477_2_2.png --output_image mixed4c_2_channel_477_2_2_2.png --octaves 8 --iter 25 --layer mixed4c --channel 477

python pb_dreamer.py --input_image mixed4c_2_channel_477_2_2_2.png --output_image mixed4c_2_channel_477_2_2_2_2.png --octaves 8 --iter 25 --layer mixed4c --channel 477

python pb_dreamer.py --input_image mixed4c_2_channel_477_2_2_2_2.png --output_image mixed4c_2_channel_477_2_2_2_2_2.png --octaves 8 --iter 25 --layer mixed4c --channel 477
```


The parameters I use seem to stay pretty consistent, so let me know if any of these work. These parameter sets come from an old laptop of mine that I used to use until it became notoriously unstable with constant crashes.


Edit:

I seem to recall running outputs through the script again multiple times to increase detail.

**Comment by ProGamerGov** [comment]
@genekogan I also just noticed that some of the lines on the image you linked to in your last comment, match this image:

![j79FllS](https://user-images.githubusercontent.com/10626398/84807005-ceb9fb80-afc3-11ea-8dcf-25bfb6e68161.jpg)


The colors seem to match as well.

So, I may have used it as a starting point. I've also in the past used random noise images from Google Images, blank images, and blank images that I've made have geometric patterns with GIMP filters.
